src/api/mc_worker_api.erl

```diff
@@ -177,44 +299,133 @@ count(Connection, Query) ->
+  case mc_utils:use_legacy_protocol() of
+      true -> 
+          command(Connection,
+                  #'query'{
+                     collection = <<"$cmd">>,
+                     selector = Command
+                    });
+      false ->
+          command(Connection, bson:fields(Command))
+  end;
+command(Connection, Command) when is_list(Command) ->
+  case mc_utils:use_legacy_protocol() of
+      true -> 
+          command(Connection, bson:document(Command));
+      false ->
+          Msg = #op_msg_command{command_doc = fix_command_obj_list(Command)},
+          {true, mc_connection_man:op_msg_raw_result(Connection, Msg)}
+  end;
+command(Connection, Command) when is_map(Command) ->
+    command(Connection, map_to_command(Command)).
+
+%% Converts map to a bson fields list. Makes sure that the command is placed
+%% first as this is a requirement from MongoDB
+map_to_command(Command) ->
+    ExistingCommands = get_existing_commands_map(),
+    IsCommand = fun(Name) -> maps:get(Name, ExistingCommands, false) end,
+    case lists:search(IsCommand, maps:keys(Command)) of
+        {value, Key} ->
+            CommandValue = maps:get(Key, Command),
+            NewCommand1 = maps:remove(Key, Command),
+            NewCommand2 = maps:to_list(NewCommand1, NewCommand1),
+            [{Key, CommandValue} | NewCommand2];
+        false ->
+            maps:to_list(Command)
+    end.
+
+fix_command_obj_list(Map) when is_map(Map) ->
+    fix_command_obj_list(maps:to_list(Map));
+fix_command_obj_list(Tuple) when is_tuple(Tuple) ->
+    fix_command_obj_list(bson:fields(Tuple));
+fix_command_obj_list(List) when is_list(List) ->
+    %% we have to try to figure out what the command field is and put it first as the command field need to go first
+    List.
 
 command(Connection, Command, _IsSlaveOk = true) ->
-  command(Connection,
-    #'query'{
-      collection = <<"$cmd">>,
-      selector = Command,
-      slaveok = true,
-      sok_overriden = true
-    });
+    case mc_utils:use_legacy_protocol() of
+        true -> 
+            command(Connection,
+                    #'query'{
+                       collection = <<"$cmd">>,
+                       selector = Command,
+                       slaveok = true,
+                       sok_overriden = true
+                      });
+        false ->
+            Command = fix_command_obj_list(Command),
+            %% slaveok seems to correspond to secondaryPreferred in the new protocol
+            CommandExtened = Command ++ [{<<"$readPreference">>, #{<<"mode">> => <<"secondaryPreferred">>}}],
+            command(Connection, CommandExtened)
+    end;
 command(Connection, Command, _IsSlaveOk = false) ->
   command(Connection, Command).
 
 %% @doc Execute MongoDB command in this thread
--spec sync_command(port(), binary(), mc_worker_api:selector(), module()) -> {boolean(), map()}.
+-spec sync_command(any(), binary(), mc_worker_api:selector(), module()) -> {boolean(), map()}.
```

Comment:
maybe define a type like this `-type socket() :: gen_tcp:socket() | ssl:sslsocket()`